### ISSUE TITLE
[Enhancement][Security] Suspend NS Client after Nightscout certificate change (certificate key pinning)

### ISSUE THREAD

**Issue by dlvoy** [MilosKozak/AndroidAPS#2440]
When connecting to Nightscout via encrypted TLS connection, first remember (and persist) what certificate it presents, and on following requests detect if certificate remains the same. If certificate has changed - suspend connections to Nightscout and warn user.

## Background
This feature is also known as certificate key pinning, as it prevents certain types of attacks when attacker take over communication channel and eavesdrop on or modify the communication. Attacker can disrupt or impersonate connection to NS (Man in the Middle attack), and inject invalid data. Potential result is disruption of loop parameters, invalid CoB/IoB calculation or further attacks (exploiting JSON parser for remote execution).

## Solution draft (to discuss):

- this feature should be use same/similar UI and framework as related #2439
- add security features to extract connection certificate details and store known certificates
- add flag that indicates that security issue was detected
- if flag is raised, block NS Client (by suspending it)
  - check checkbox "suspended" in NSClient
  - make checkbox not uncheckable (grayed out)
  - add small (i) icon next to checkbox, clicking it leads to confirmation dialog
  - add global warning (red notification) where dismiss leads to confirmation dialog
  - clicking on info icon will pop up security confirmation dialog:
    > AAPS detected that Nightscout server changed it's certificate. It may be due to server or certificate update / renewal (which happens once in a while and is OK) but can also indicate attack on your network or Nightscout instance!
    > In case of attack: your API secret and private medical data may be exposed and potentially allow attacker to disrupt your current loop status, which may be harmful or even lethal!
    > 
    > Please check your network connection and Nightscout to be ABSOLUTELY sure that your connection to Nightscout is secure. You may open Nightscout in web browser, click on padlock icon and check certificate details
    > 
    >  ☐   I have checked my Nightscout and confirmed that everything is OK, and want allow connection to it.
    > _checking that checkbox make ALLOW button enabled_
    >
    > [ Trust Nightscout ] _(disabled)_                        [UNDERSTOOD]
  - checking checkbox and clicking trust button will update pinned certificate for given server
- there should be configuration option to allow/disable that check (enabled by default)
  
## Variants / Extensions / Ideas
- list of trusted certificates
- part of (new) `Security` plugin - showing list of allowed WiFi and trusted Nightscout server certificates

**Comment by asquelt** [comment]
might be ineffective for users who use LE certificates, as they change often. securing a communication, despite more complex to implement, would be more effective than trying to secure insecure communication medium.

**Comment by dlvoy** [comment]
i agree that additional two-way securing of communication between NS and AAPS would be better, but it would require non trivial changes to NS, having user configuring it properly on NS that will support it - definitely story for the future, but wont happen i near future

**TL;DR** something is better than nothing, and there is no cure against human nature

about mentioned scenarios:
* for Lets Encrypt and other - if we pin on intermediate cert too it can be still useful, since they do not change so often. At end, only issue would be prompt from time to time, that make awareness to the certificate change - and user can still invest and pay for more permanent cert. This issue affects users that self-host NS, and they usually know what they are doing. Most others use services like Heroku which has commercial grade cert.
* for corporate WiFi - user should know already that network is under WAF, it is usually under consent or in work regulations / BYOD policy / contract / state law etc. At end - it is still **choice** of user to allow his private data go trough supervised channel, this feature only raises awareness of that dealings
* for sure most users wont read, but spectrum is wide and at other end some will even build their own forks of AAPS to bypass any clever mechanism that was built for their own security. We cannot foolproof against recklessness or stupidity, only put some features and educate users about why and how they can protect them (a little bit more). It is why i suggested such long explanation in popup, but also option to bypass it. Assumption of AAPS is that this is DIY system and you roughly know what you are doing. Barrier to entry is purposefully high, and you know and take all involved risks.

**Comment by asquelt** [comment]
my general opinion on this is that it's viable way to mitigate a risk of MITM and trying to hack AAPS remotely without a need to alter NS codebase.

approperiate entry in documentation about remoting with AAPS should be introduced to make users aware of risks involved in remoting.

it should be considered whether NS should be required:
* in Objectives
* with default of ns_upload_only => false

preferably AAPS should be disconnected, or write-only by default.

existing users should be made aware of change of defaults (if any).

**it should be prioritized as a low hanging fruit security solution** before long-term fix is made utilizing mutual authentication mechanism between two parties.

**Comment by dlvoy** [comment]
feel free to move paragraph above to separate issue for that matter, you are also welcome to discuss that on gitter to agree with other maintainers if that is preferable configuration for most users and should be default - and after gaining support for that concept - you will only need to write PRs for app code and documentation and get them trough review, test and into release

**Comment by twain47** [comment]
Could you give an example of a practical MITM attack on TLS that doesn't involve either a corporate network installing their own certificate (which the user should be aware of) or compromising the server/DNS/email in order to obtain a certificate for the relevant domain (which probably makes it irrelevant because at that point it is all compromised anyway)?

I'm not saying what you are suggesting wouldn't enhance security (belt and braces are not a bad thing in security), but what are you trying to mitigate?

**Comment by dlvoy** [comment]
Duno if it was question for me or @asquelt. Personally i asses risk for MITM on NS connection as very low and impact as moderate, since AAPS has many other security features that prevent overdosing even with maliciously crafted and injected profile/IoB/CoB data. 

Some users will use dedicated phone for AAPS (as it is encouraged), with upcoming better local profile dependency on NS become weaker, but it is still useful tool and two way sync is sometimes just too convenient to opt-out.

Despite low probability of exploiting that weakness i would still want to have it in place, and volunteer to implement it. It is done industry-wide for less mission-critical apps. For sure it will allow some more anxious users to sleep better, and improve overall AAPS security with yet one countermeasure - more for PR and as emergency backup plan - but hey, that matters to!

**Comment by asquelt** [comment]
a practical attack could be implemented using rouge access point configured in such a way, that it is broadcasting SSID that phone has connected in the past. once connected, rouge actor may relay all internet traffic, including planting their own DNS entries and their own LE certificates connected to their own modified version of NS. however, I believe this is not the only possible attack, just the one that can be simply reproduced and demoed to prove the point.

like Dominik wrote, the whole communication in AAPS is not up to the current security standards, where there's common agreement on strong mutual authentication and distrust in communications channels underneath.

a way less critical applications, like your cloud apps, are using better authentication, and there is no life hazard involved with using them. personally I think there are majority of children using AAPS connected systems and it only takes time that some of them will exhibit curiosity about how things work underneath the AAPS system and try to exploit them. they might be skilled enough and not mature enough to put their peers to risk while performing such an experiment.

**Comment by twain47** [comment]
@asquelt you can't obtain a LE certificate without already having control of the target DNS/server.

Systems like banks trust this level of security without issue and with a much higher payoff if someone is able to compromise them.

**Comment by asquelt** [comment]
there is evidence in history on compromised CAs or buggy client implementations that don't even look in CN or the chain of trust (i'm not sure how AAPS approaches security of libraries being used). the margin for exploitation is very broad, thus there's common consensus to not to trust single security mechanism, but otherwise tier them together. it's much less likely to evade them all at the same time.

banks run their model on insurance, whereas AAPS should run theirs on users safety and well being. i think we shouldn't base on banks threat model here.

**Comment by AdrianLxM** [comment]
> users that connect to corporate wifi, where SSL certificate substitution is performed
@asquelt @dlvoy As you seem to have expertise here: Wouldn't that mean that on the phone there would have to be user-added certificates?
On newer Android versions by default only system-provided certificates are accepted for apps. We could exclude user-provided ones for older versions (Android 6 and lower) as well.

**Comment by asquelt** [comment]
it is possible in android to import additional CA. user may be asked (by network owner policy) or tricked (by social engineering attempt) into onboarding one.

**Comment by dlvoy** [comment]
For devices provisioned for corporate network - Android for work API will be used to configure phone as BYOD device or partially fully managed or partial (like: with switchable work profile). System policy administrator / usually with some EMM / MAM tools, will manage also root CA store - usually to provision own root or intermediate certs. Of course user -usually- will be fully aware for this, for example with work profile CA will be only affecting apps in that domain, not "private" apps...

... but my concern is that for some other (most?) users all that what we discuss here is black magic, I've seen people trying to build AAPS on employer-provided PC and wondering why it is not working (while there was firewall or management software blocking it)

**Comment by twain47** [comment]
@dlvoy This is also my concern (that this is black magic to most users). Certificate pinning could result in a warning (or even worse NS being blocked) every time LE updated the certificate - so every 2 months.  There are far more users who have to deal with a LE updated certificate than an invasive corporate firewall.

Frankly this sounds more a case for a discussion with the company IT. I'd be making them aware they were snooping on medical data and watch their faces to white.
